l1-contracts/src/governance/libraries/BLS.sol

```diff
@@ -0,0 +1,374 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2024 Aztec Labs.
+pragma solidity >=0.8.27;
+
+// Inspiration: https://gist.github.com/kobigurk/257c1783ddf556e330f31ed57febc1d9
+// Helpers taken from: https://github.com/thehubbleproject/hubble-contracts/blob/master/contracts/libs/BLS.sol
+
+import {ModexpInverse, ModexpSqrt} from "./ModExp.sol";
+
+library BLS {
+  uint256 public constant FIELD_SIZE =
+    21888242871839275222246405745257275088696311157297823662689037894645226208583;
+
+  uint256 public constant CURVE_ORDER =
+    21888242871839275222246405745257275088548364400416034343698204186575808495617;
+
+  // G1 Generator
+  uint256 public constant G1_X = 1;
+  uint256 public constant G1_Y = 2;
+
+  // Negated generator of G2
+  uint256 public constant NEG_G2_X1 =
+    11559732032986387107991004021392285783925812861821192530917403151452391805634;
+  uint256 public constant NEG_G2_X0 =
+    10857046999023057135944570762232829481370756359578518086990519993285655852781;
+  uint256 public constant NEG_G2_Y1 =
+    17805874995975841540914202342111839520379459829704422454583296818431106115052;
+  uint256 public constant NEG_G2_Y0 =
+    13392588948715843804641432497768002650278120570034223513918757245338268106653;
+
+  bytes32 public constant STAKING_DOMAIN_SEPARATOR = bytes32("AZTEC_BLS_POP_BN254_FT_V1");
+
+  // sqrt(-3)
+  uint256 private constant Z0 = 0x0000000000000000b3c4d79d41a91759a9e4c7e359b6b89eaec68e62effffffd;
+  // (sqrt(-3) - 1)  / 2
+  uint256 private constant Z1 = 0x000000000000000059e26bcea0d48bacd4f263f1acdb5c4f5763473177fffffe;
+  uint256 private constant T24 = 0x1000000000000000000000000000000000000000000000000;
+  uint256 private constant MASK24 = 0xffffffffffffffffffffffffffffffffffffffffffffffff;
+
+  error pk1Zero();
+  error pk2Zero();
+  error signatureZero();
+  error addPointFail();
+  error mulPointFail();
+  error gammaZero();
+  error pairingFail();
+  error valueOutOfRange(uint256 value, uint256 min, uint256 max);
+  error noPointFound();
+
+  /**
+   * @notice Register a BLS signature key for the caller.
+   *      Each unique key can only be registered once across all validators.
+   *
+   * @param pk1 The G1 point of the BLS public key (x, y coordinates)
+   * @param pk2 The G2 point of the BLS public key (x1, x0, y1, y0 coordinates)
+   * @param signature The G1 point that acts as a proof of possession of the private keys corresponding to pk1 and pk2
+   */
```

Comment:
```suggestion
   * Ultimately, we want to check:
   * - That the caller knows the secret key of pk2 (to prevent rogue-key attacks)
   * - That pk1 and pk2 have the same secret key (as an optimisation)
   * Registering two public keys is an optimisation: It means we can do G1-only operations
   * at the time of verifying a signature, which is much cheaper than G2 operations.
   *
   * In this function, we check:
   * e(signature + gamma * pk1, -G2) * e(hashToPoint(pk1) + gamma * G1, pk2) == 1
   * 
   * Which is effectively a check that:
   * e(signature, G2) == e(hashToPoint(pk1), pk2) // a BLS signature over msg = pk1, to prove knowledge of the sk.
   * e(pk1, G2) == e(G1, pk2) // a demonstration that pk1 and pk2 have the same sk. 
   */
```